hashicorp/terraform-provider-tfe
# Add project level auto destroy

## Description

Add support in the tfe provider for the project level ephemeral workspace settings.

_Remember to:_


## Testing plan

1. Set up a workspace inside a project like such
```
resource "tfe_project" "test" {
  organization = var.organization
  name = "test-project"
  auto_destroy_activity_duration = "9h"
}

resource "tfe_workspace" "overriding_workspace" {
  organization = var.organization
  name = "test_workspace"
  project_id = tfe_project.test.id

  vcs_repo {
    identifier = var.repo
    branch = var.branch
    oauth_token_id = tfe_oauth_client.github.oauth_token_id
  }
}
```
2. The workspace should pick up the setting from the project. You can verify this by going to `/app/<orgname>/workspaces/<workspace_name>/settings/delete`
3. Set up an overriding workspace
```
resource "tfe_project" "test" {
  organization = var.organization
  name = "test-project"
  auto_destroy_activity_duration = "9h"
}

resource "tfe_workspace" "overriding_workspace" {
  organization = var.organization
  name = "test_workspace"
  project_id = tfe_project.test.id
  auto_destroy_activity_duration = "10d"
  inherits_project_auto_destroy = false

  vcs_repo {
    identifier = var.repo
    branch = var.branch
    oauth_token_id = tfe_oauth_client.github.oauth_token_id
  }
}
```
4. The workspace should override the setting from the project.

## External links

_Include any links here that might be helpful for people reviewing your PR. If there are none, feel free to delete this section._

- [go-tfe PR](https://github.com/hashicorp/go-tfe/pull/1011)

## Output from acceptance tests

_Please run applicable acceptance tests locally and include the output here. See [testing.md](https://github.com/hashicorp/terraform-provider-tfe/blob/main/docs/testing.md) to learn how to run acceptance tests._

Project Resource

Workspace Resource

Project Data source

```
TESTARGS="-run TestAccTFEProjectDataSource_*" envchain tf-provider make testacc          $! simonxmh/add_project_level_auto_destroy 
TF_ACC=1 TF_LOG_SDK_PROTO=OFF go test $(go list ./... |grep -v 'vendor') -v -run TestAccTFEProjectDataSource_* -timeout 15m
?       github.com/hashicorp/terraform-provider-tfe     [no test files]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-tfe/internal/client     0.293s [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-tfe/internal/logging    0.470s [no tests to run]
?       github.com/hashicorp/terraform-provider-tfe/internal/provider/validators        [no test files]
?       github.com/hashicorp/terraform-provider-tfe/version     [no test files]
=== RUN   TestAccTFEProjectDataSource_basic
--- PASS: TestAccTFEProjectDataSource_basic (9.52s)
=== RUN   TestAccTFEProjectDataSource_caseInsensitive
--- PASS: TestAccTFEProjectDataSource_caseInsensitive (7.30s)
=== RUN   TestAccTFEProjectDataSource_basicWithAutoDestroy
--- PASS: TestAccTFEProjectDataSource_basicWithAutoDestroy (9.16s)
PASS
ok      github.com/hashicorp/terraform-provider-tfe/internal/provider   26.446s
```

Workspace Data source

```
TESTARGS="-run TestAccTFEWorkspaceDataSource_*" envchain tf-provider make testacc        
TF_ACC=1 TF_LOG_SDK_PROTO=OFF go test $(go list ./... |grep -v 'vendor') -v -run TestAccTFEWorkspaceDataSource_* -timeout 15m
?       github.com/hashicorp/terraform-provider-tfe     [no test files]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-tfe/internal/client     0.471s [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-tfe/internal/logging    (cached) [no tests to run]
?       github.com/hashicorp/terraform-provider-tfe/internal/provider/validators        [no test files]
?       github.com/hashicorp/terraform-provider-tfe/version     [no test files]
=== RUN   TestAccTFEWorkspaceDataSource_remoteStateConsumers
--- PASS: TestAccTFEWorkspaceDataSource_remoteStateConsumers (11.11s)
=== RUN   TestAccTFEWorkspaceDataSource_basic
--- PASS: TestAccTFEWorkspaceDataSource_basic (6.16s)
=== RUN   TestAccTFEWorkspaceDataSourceWithTriggerPatterns
--- PASS: TestAccTFEWorkspaceDataSourceWithTriggerPatterns (4.11s)
=== RUN   TestAccTFEWorkspaceDataSource_readAutoDestroyAt
--- PASS: TestAccTFEWorkspaceDataSource_readAutoDestroyAt (10.25s)
=== RUN   TestAccTFEWorkspaceDataSource_readAutoDestroyDuration
--- PASS: TestAccTFEWorkspaceDataSource_readAutoDestroyDuration (10.36s)
=== RUN   TestAccTFEWorkspaceDataSource_readProjectIDDefault
--- PASS: TestAccTFEWorkspaceDataSource_readProjectIDDefault (6.05s)
=== RUN   TestAccTFEWorkspaceDataSource_readProjectIDNonDefault
--- PASS: TestAccTFEWorkspaceDataSource_readProjectIDNonDefault (7.61s)
PASS
ok      github.com/hashicorp/terraform-provider-tfe/internal/provider   56.248s
TESTARGS="-run TestAccTFEWorkspaceDataSource_*" envchain tf-provider make   7.37s user 4.54s system 20% cpu 59.017 total

...
```
